artifactory/commands/transferfiles/delayedartifactshandler.go

```diff
@@ -0,0 +1,209 @@
+package transferfiles
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/jfrog/jfrog-client-go/utils/errorutils"
+	"github.com/jfrog/jfrog-client-go/utils/io/content"
+	"github.com/jfrog/jfrog-client-go/utils/log"
+	"os"
+	"path"
+	"path/filepath"
+)
+
+const (
+	// Max delayedArtifacts that will be written in a file
+	maxDelayedArtifactsInFile = 50000
+)
+
+// TransferDelayedArtifactsMng managing multi threads writing delayed deployment artifacts in order to keep the order of deployment.
+type TransferDelayedArtifactsMng struct {
+	// All go routines will write delayedArtifacts to this channel
+	delayedArtifactsChannel chan FileRepresentation
+
+	// Files containing delayed artifacts to upload later on.
+	filesToConsume []string
+	deployedWriter delayedArtifactWriter
+}
+
+type delayedArtifactWriter struct {
+	writer               *content.ContentWriter
+	delayedArtifactCount int
+	// In case we have multiple delayedArtifacts files - we index them
+	fileIndex int
+}
+
+func (mng *TransferDelayedArtifactsMng) initDelayedArtifactWriter() (err error) {
+	// Init the content writer which responsible for writing delayed artifacts - That means delayed artifacts that we should upload later, in a certain order.
+	writer, err := mng.newContentWriter()
+	if err != nil {
+		return err
+	}
+	mng.deployedWriter = delayedArtifactWriter{writer: writer, fileIndex: 0}
+	return nil
+}
+
+// Creates a manager for the files transferring process.
+func newTransferDelayedArtifactsToFile(delayedArtifactsChannel chan FileRepresentation) (*TransferDelayedArtifactsMng, error) {
+	mng := TransferDelayedArtifactsMng{delayedArtifactsChannel: delayedArtifactsChannel}
+	err := mng.initDelayedArtifactWriter()
+	return &mng, err
+}
+
+func (mng *TransferDelayedArtifactsMng) start() error {
+	for file := range mng.delayedArtifactsChannel {
+		err := mng.writeDelayedArtifactContent(file)
+		if err != nil {
+			return err
+		}
+	}
+	return mng.closeWriter()
```

Comment:
As mentioned in a previos comment, this should happen ina defer, and also inside the dame function which created the writer.